Quotient/e2ee/cryptoutils.cpp

```diff
@@ -0,0 +1,267 @@
+        return ERR_get_error();
+    }
+
+    status = EVP_EncryptInit_ex(ctx, EVP_aes_256_ctr(), nullptr, reinterpret_cast<const unsigned char*>(key.data()), reinterpret_cast<const unsigned char*>(iv.data()));
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        EVP_CIPHER_CTX_free(ctx);
+        return ERR_get_error();
+    }
+
+    status = EVP_EncryptUpdate(ctx, reinterpret_cast<unsigned char*>(encrypted.data()), &length, reinterpret_cast<const unsigned char *>(&plaintext.data()[0]), (int) plaintext.size());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        EVP_CIPHER_CTX_free(ctx);
+        return ERR_get_error();
+    }
+
+    ciphertextLength = length;
+    status = EVP_EncryptFinal_ex(ctx, reinterpret_cast<unsigned char*>(encrypted.data()) + length, &length);
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        EVP_CIPHER_CTX_free(ctx);
+        return ERR_get_error();
+    }
+
+    ciphertextLength += length;
+    encrypted.resize(ciphertextLength);
+    EVP_CIPHER_CTX_free(ctx);
+    return encrypted;
+}
+
+SslExpected<HkdfKeys> Quotient::hkdfSha256(const QByteArray& key,
+                                           const QByteArray& salt,
+                                           const QByteArray& info)
+{
+    QByteArray result(64, u'\0');
+    auto context = EVP_PKEY_CTX_new_id(EVP_PKEY_HKDF, nullptr);
+
+    int status = EVP_PKEY_derive_init(context);
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    status = EVP_PKEY_CTX_set_hkdf_md(context, EVP_sha256());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    status = EVP_PKEY_CTX_set1_hkdf_salt(context, reinterpret_cast<const unsigned char *>(salt.data()), salt.size());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    status = EVP_PKEY_CTX_set1_hkdf_key(context, reinterpret_cast<const unsigned char *>(key.data()), key.size());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    status = EVP_PKEY_CTX_add1_hkdf_info(context, reinterpret_cast<const unsigned char *>(info.data()), info.size());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    std::size_t outputLength = result.size();
+    status = EVP_PKEY_derive(context, reinterpret_cast<unsigned char *>(result.data()), &outputLength);
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    EVP_PKEY_CTX_free(context);
+
+    if (outputLength != 64) {
```

Comment:
In this case, I guess we could put a `Q_ASSERT` but otherwise return the buffer, or, harsher, just `std::terminate()`? I don't see how the calling code can recover from this.

Reply:
The caller could just... cancel the crypto stuff they're doing and... continue running; in practice, there's no reason for this to happen other than a bug in libquotient (or openssl), so terminating would be the best thing to do IMO

Reply:
As far as I gather, it is not a given that the output length should be exactly 64 bytes for any inputs; for cryptoutils, I decided to still return an error code from cryptoutils (fortunately, OpenSSL's error code space _is_ extensible, unlike Olm's) and later see if more assertions/terminations should be introduced up the stack.